src/tools/illink/src/linker/Linker/TypeMapInfo.cs

```diff
@@ -329,22 +332,29 @@ void AnnotateMethods (MethodDefinition @base, MethodDefinition @override, Interf
 		/// <param name="implOfInterface">
 		/// The InterfaceImplementation on <paramref name="type"/> that points to the DeclaringType of <paramref name="interfaceMethod"/>.
 		/// </param>
-		void FindAndAddDefaultInterfaceImplementations (TypeDefinition typeThatImplementsInterface, TypeDefinition typeThatMayHaveDIM, MethodDefinition interfaceMethodToBeImplemented, InterfaceImplementation originalInterfaceImpl)
+		void FindAndAddDefaultInterfaceImplementations (RuntimeInterfaceImplementation originalInterfaceImpl, MethodReference inflatedInterfaceReference, MethodDefinition interfaceMethodToBeImplemented)
 		{
+			var typeThatImplementsInterface = originalInterfaceImpl.Implementor;
 			// Go over all interfaces, trying to find a method that is an explicit MethodImpl of the
 			// interface method in question.
+			var hasInterfaces = interfaces.TryGetValue (typeThatImplementsInterface, out var runtimeIfaces);
+			if (!hasInterfaces)
+				return;
 
-			foreach (var interfaceImpl in typeThatMayHaveDIM.Interfaces) {
-				var potentialImplInterface = context.TryResolve (interfaceImpl.InterfaceType);
-				if (potentialImplInterface == null)
+			bool findAllPossibleImplementations = inflatedInterfaceReference.DeclaringType is GenericInstanceType;
+			foreach (var interfaceImpl in runtimeIfaces!) {
+				var potentialImplInterface = interfaceImpl.InterfaceTypeDefinition;
+				if (potentialImplInterface is null)
 					continue;
 
+				// We go by MethodDefinition here and may miss the correct implementation by exiting early after finding one.
+				// To correct for this, we'll look for all possible implementations and add them all if the inflatedIntefaceMethodReference is a genericInstance
+				// Otherwise, we're good to stop looking higher in the hierarchy after the first. Note we still need to look at other branches of the hierarchy to find any diamond cases.
```

Comment:
I'm alright to separate this out, but FWIW it would only remove the `findAllPossibleImplementations` references on line 346 and 371 (in source code - there would be more test changes).

Reply:
Pulled those changes out in https://github.com/dotnet/runtime/pull/103317/commits/bb308c7e99e1576625bcba9052ad6d51a4f76bb9